kuma/spam/akismet.py

```diff
@@ -0,0 +1,215 @@
+import sys
+import urlparse
+
+import newrelic.agent
+from constance import config
+from django.conf import settings
+from requests import Session
+from requests.packages.urllib3.util import Retry
+from requests.adapters import HTTPAdapter
+from requests.exceptions import RequestException
+
+
+class AkismetError(Exception):
+    """
+    A custom exception that takes the requests' response, its status code
+    and an optional debug help directly from Akismet.
+    """
+    def __init__(self, response, status_code, debug_help):
+        self.response = response
+        self.status_code = status_code
+        self.debug_help = debug_help
+
+    def __str__(self):
+        return ('%s response, debug help: %s, full response: %s' %
+                (self.status_code, self.debug_help, self.response.text))
+
+
+class Akismet(object):
+    """
+    The Akismet client implementation wrapping a request session and
+    having API specific methods for easy spam and ham handling.
+    """
+    # Warning: This is the documented success response from the Akismet API
+    # Do not translate! Info: http://akismet.com/development/api/#submit-spam
+    submission_success = 'Thanks for making the web a better place.'
+
+    def __init__(self):
+        self.domain = settings.DOMAIN
+        self.session = Session()
+        self.adapter = HTTPAdapter(
+            max_retries=Retry(
+                # number of total retries
+                total=5,
+                # retry once in case we can't connect to Akismet
+                connect=1,
+                # retry once in case we can't read the response from Akismet
+                read=1,
+                # retry once in case we're redirect by Akismet
+                redirect=1,
+                # definitely retry if Akismet is unwell
+                status_forcelist=[500, 503])
+        )
+        self.session.mount('http://', self.adapter)
+        self.session.mount('https://', self.adapter)
+        self._verified = None
+
+    @property
+    def key(self):
+        """
+        The Akismet API key, implemented as a property so the client
+        can be instanciated as a class attribute and only later receive
+        the key from constance.
+        """
+        return config.AKISMET_KEY
+
+    @property
+    def ready(self):
+        """"
+        Returns whether this client is usable by verifying the API key.
+        """
+        if self._verified is None:
+            self._verified = self.verify_key()
+            if self._verified is None:
+                return False
+        return self._verified
+
```

Comment:
The property needs to be able to evaluated multiple time, e.g. when the key verification doesn't work the first time for some reasons (e.g. connectivity issue) the result will be `None` and by that tricker a retry of the verification on the next request (form instance).